foundation-model-stack/fms-hf-tuning
# test: add arrow datasets and arrow unit tests

### Description of the change
Adding changes to enable loading of Arrow Dataset files by datasets.load_dataset

Datasets were converted from json using the following script:
```
import pyarrow as pa
import pyarrow.json as pajson
import json

json_file = <file>
table = pajson.read_json(json_file)

output_file = <output file>
with pa.OSFile(output_file, 'wb') as sink:
    with pa.ipc.new_file(sink, table.schema) as writer:
        writer.write(table)
```

Using this script we can take a look at the new datasets added:
```
import pyarrow as pa
import pyarrow.ipc as ipc

arrow_file = "tests/artifacts/testdata/twitter_complaints_small.arrow"

with pa.OSFile(arrow_file, 'rb') as source:
    reader = ipc.open_file(source)
    table = reader.read_all()

print(table)

df = table.to_pandas()
print(df)
```

pretokenized:
![image](https://github.com/user-attachments/assets/6bafe1fd-8fd3-49f9-8e1b-049d383b9390)

input_output:
![image](https://github.com/user-attachments/assets/7e959ce2-1fdb-4112-aa8c-e92da9a9057b)

small:
![image](https://github.com/user-attachments/assets/3d90059f-72b9-49b7-bf3f-80be282cd551)


### Related issue number
Following PR #401 


### How to verify the PR
Run
```
python tuning/sft_trainer.py  \
--model_name_or_path Maykeye/TinyLLama-v0  \
--training_data_path tests/artifacts/testdata/twitter_complaints_input_output.arrow \
--output_dir outputs/full-tuning  \
--num_train_epochs 5  \
--per_device_train_batch_size 2  \
--gradient_accumulation_steps 1  \
--learning_rate 1e-5  \
--use_flash_attn false \
--torch_dtype "float32"
```

`tox -e py`

To see end to end build and results on llama3 8b, see [Travis CI build](https://v3.travis.ibm.com/github/ai-foundation/fmaas-tuning-docs/builds/30735339) (click on jobs and look through logs to see training, inference, response)

### Was the PR tested
